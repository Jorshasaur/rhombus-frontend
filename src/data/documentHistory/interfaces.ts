export interface DocumentHistoryState {
    showDocumentHistory: boolean
}
