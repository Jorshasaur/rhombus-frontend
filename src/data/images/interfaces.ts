export interface ActiveImageInfo {
    fileName: string
    author: string
    createdAt?: string
    url?: string
    width?: number
    height?: number
}

export interface ImageCarouselImage {
    url: string
    fileName: string
    width?: number
    height?: number
}

export interface ImageCarouselData {
    left: ImageCarouselImage
    middle: ImageCarouselImage
    right: ImageCarouselImage
}

export interface ImageDataStore {
    id: string
    assetId?: string
    authorId: string
    url: string
    createdAt?: string
    width?: number
    height?: number
}

export interface ImagesStoreState {
    activeImage: string
    imageDatas: { [imageId: string]: ImageDataStore }
    imageList: string[]
}
