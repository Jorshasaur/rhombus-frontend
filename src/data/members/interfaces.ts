export interface NavigationMember {
    avatarURL: string
    id: number
    index: number
    name: string
    useAvatarURL: boolean
}
