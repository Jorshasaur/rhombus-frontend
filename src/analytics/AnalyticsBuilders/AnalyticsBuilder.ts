import analytics from '../analytics'

interface EventProperties {
    [key: string]: any
}

export default abstract class AnalyticsBuilder {
    protected abstract eventName: string

    private _properties: EventProperties = {}

    withProperty(key: string, value: any) {
        this._properties[key] = value
        return this
    }

    withProperties(properties: { [key: string]: any }) {
        this._properties = {
            ...this._properties,
            ...properties
        }
        return this
    }

    track() {
        analytics.track(this.eventName, this._properties)
    }

    public getEventName() {
        return this.eventName
    }

    public readProperties() {
        return Object.assign({}, this._properties)
    }
}
