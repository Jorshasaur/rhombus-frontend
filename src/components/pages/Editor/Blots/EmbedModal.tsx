import React from 'react'
import styles from './EmbedModal.module.css'
import { Close } from '@invisionapp/helios/icons'
import ReactDOM from 'react-dom'
import { ThemeProvider } from 'styled-components'
import theme from '@invisionapp/helios/css/theme'
import { ErrorBoundary } from '../../../../bugsnag'

function invariant(predicate: boolean, message: string) {
    if (process.env.NODE_ENV !== 'production') {
        if (!predicate) {
            console.error(message)
        }
    }
}

export interface EmbedModalProps {
    actionArea?: React.ReactNode
    children: React.ReactNode
    onHide: (event: React.MouseEvent<HTMLAnchorElement, MouseEvent>) => void
}

interface State {
    isMounted: boolean
    isShown: boolean
}

/**
 * This is not a normal React component, but a static singleton that must be used
 * via its public static methods. It is used by mounting it with specific data, and
 * then even when the "parent" component unmounts, the EmbedModal remains open and
 * unaffected until it is explicitly unmounted.
 *
 * In a Quill context, blots are often deleted and inserted which means any child components
 * are deleted and recreated (not just updated) all the time, and the modal needs to persist
 * those changes.
 *
 * Here is an example:
 * ```tsx
 * class NeatComponent extends React.Component {
 *   componentDidMount() {
 *     EmbedModal.mount()
 *   }
 *
 *   getEmbedModalData = (): EmbedModalProps => ({
 *     onHide: (event) => this.handleModalClose(event),
 *     actionArea: <a>Open in new tab...</a>,
 *     children: <iframe src='http://google.com' />
 *   })
 *
 *   handleModalOpen = () => {
 *     EmbedModal.show(this.getEmbedModalData())
 *   }
 *
 *   handleModalClose = () => {
 *     EmbedModal.unmount()
 *   }
 *
 *   render() {
 *     return <button onClick={this.handleModalOpen}>open modal</button>
 *   }
 * }
 * ```
 */
export class EmbedModal extends React.Component<Partial<EmbedModalProps>> {
    static defaultProps = {
        onHide: () => undefined
    }
    private static root: HTMLDivElement = document.createElement('div')
    private static state: State = {
        isMounted: false,
        isShown: false
    }
    private static props: Partial<EmbedModalProps> = {}

    /**
     * Initialize the modal and prepare it to be shown. This method does *not*
     * actually show the modal, but just attaches the content to the DOM ahead of time
     * so that it is animatable when it opens. This only really needs to be done once.
     */
    public static mount() {
        if (!EmbedModal.state.isMounted) {
            EmbedModal.setState({
                isMounted: true,
                isShown: false
            })
            EmbedModal.render()
        }
    }

    /**
     * Once the modal is mounted, call `EmbedModal.show()` to make it open. This
     * will typically happen as a result of a button click or something similar.
     *
     * @param props the content to hydrate the modal with
     */
    public static show(props: EmbedModalProps) {
        invariant(
            EmbedModal.state.isMounted,
            `Warning: you are trying to call EmbedModal.show, but EmbedModal.mount must be called first and it wasn't. The modal will not work correctly before it is mounted.`
        )

        if (EmbedModal.state.isMounted) {
            EmbedModal.props = props
            EmbedModal.setState({ isShown: true })
            EmbedModal.render()
        }
    }

    /**
     * A utility method to determine whether the modal is currently open or not.
     */
    public static isShown() {
        return EmbedModal.state.isShown
    }

    /**
     * For when you need to update any props since the modal was mounted
     * @param props any props that need to be updated
     */
    public static update(props: Partial<EmbedModalProps>) {
        EmbedModal.props = { ...EmbedModal.props, ...props }
    }

    /**
     * This method will close the modal. *Note*: this does not remove the nodes from
     * the DOM, because otherwise you would have to mount it again before you could open it.
     */
    public static hide() {
        if (EmbedModal.state.isMounted && EmbedModal.isShown()) {
            EmbedModal.setState({ isShown: false })
            EmbedModal.render()
        }
    }

    /**
     * Call `unmount` to completely remove the modal content from the DOM. *Note*: You probably
     * do _not_ want to call this in the parent's `componentWillUnmount` because then it will be
     * removed from the DOM when a Quill blot gets deleted/inserted.
     */
    public static unmount() {
        if (EmbedModal.state.isMounted) {
            EmbedModal.setState({ isMounted: false })
            EmbedModal.hide()
            setTimeout(() => {
                ReactDOM.unmountComponentAtNode(EmbedModal.root)
            }, 500)
        }
    }

    /**
     * This private method acts similarly to `React.createPortal` except it decouples
     * the rendered modal from the parent's lifecycle.
     */
    private static render() {
        ReactDOM.render(
            <ErrorBoundary>
                <EmbedModal {...EmbedModal.props} />
            </ErrorBoundary>,
            EmbedModal.root
        )
    }

    private static setState(state: Partial<State>) {
        EmbedModal.state = { ...EmbedModal.state, ...state }
    }

    private constructor(props: EmbedModalProps) {
        super(props)
        if (!document.querySelector('#embed-modal')) {
            EmbedModal.root.id = 'embed-modal'
            document.body.appendChild(EmbedModal.root)
        }
    }

    render() {
        return (
            <ThemeProvider theme={theme}>
                <div
                    className={
                        EmbedModal.isShown()
                            ? styles.backdropVisible
                            : styles.backdrop
                    }>
                    <div className={styles.modal}>
                        <div className={styles.closeButtonContainer}>
                            <a
                                onClick={this.props.onHide}
                                className={styles.closeButton}
                                data-testid="embed-modal__close-button">
                                <Close fill="white" size={24} />
                            </a>
                        </div>
                        <div className={styles.actionArea}>
                            {this.props.actionArea}
                        </div>
                        <div
                            className={
                                EmbedModal.isShown()
                                    ? styles.bodyVisible
                                    : styles.body
                            }>
                            {EmbedModal.isShown() && this.props.children}
                        </div>
                    </div>
                </div>
            </ThemeProvider>
        )
    }
}
