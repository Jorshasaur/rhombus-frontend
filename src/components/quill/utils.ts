import { DeltaStatic } from 'quill-delta'
import Quill from 'quill/core'
import { CommentMention } from '../../data/services/PagesApiService'
import { Blot } from 'parchment/dist/src/blot/abstract/blot'

interface CommentTextAndMentions {
    text: string
    source: string
    mentions: CommentMention[]
}

export function getBounds(bounds: ClientRect, containerBounds: ClientRect) {
    return {
        top: bounds.top - containerBounds.top,
        right: bounds.right - containerBounds.left,
        bottom: bounds.bottom - containerBounds.top,
        left: bounds.left - containerBounds.left,
        width: bounds.width,
        height: bounds.height
    }
}

export function getCachedContents(quill: Quill) {
    return quill.editor.delta
}

export function getTextAndMentions(delta: DeltaStatic) {
    const res = { text: '', source: '', mentions: [] }

    if (!delta) {
        return res
    }

    return delta.reduce((ret: CommentTextAndMentions, op) => {
        if (op.insert != null) {
            if (typeof op.insert === 'string') {
                ret.text += op.insert
                ret.source += op.insert
            } else if (op.insert.mention) {
                const { name, userId } = op.insert.mention
                ret.text += `@${name}`
                ret.source += `<@U${userId}>`
                ret.mentions.push({
                    token: name,
                    userId
                })
            }
        }
        return ret
    }, res)
}

export function preloadImage(imageUrl: string) {
    const preload = new Image()
    preload.src = imageUrl
}

export function getBlotOffset(quill: Quill, blot: Blot) {
    let offset = blot.offset()

    while (blot.parent !== quill.scroll) {
        blot = blot.parent
        offset += blot.offset()
    }
    return offset
}

// It's odd but the line length for the added last line is actually 1 because of the
// invisible line break character added to create the line.
export function isBlankLine(lineLength: number, lineContent: string) {
    return lineLength === 1 && lineContent === ''
}
