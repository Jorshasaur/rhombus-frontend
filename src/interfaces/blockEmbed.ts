import FileEmbed from '../components/pages/Editor/Blots/FileEmbed'
import DefaultEmbedTypes from '../components/quill/blots/DefaultEmbedTypes'
import { BlotSize } from './blotSize'

export type BlockEmbedService =
    | keyof typeof DefaultEmbedTypes
    | ReturnType<FileEmbed['_getFileTypeClass']>

export interface BlockEmbedValue {
    authorId: string
    createdAt: string
    dataUrl?: string
    embed?: any
    embedData: any
    originalLink?: string
    service: BlockEmbedService
    size?: BlotSize
    type?: string
    unviewable?: boolean
    uuid: string
    version: number
}

export interface BlockEmbedProps {
    authorId: string
    authorName?: string
    createdAt?: string
    embedData?: {}
    key: string
    originalLink?: string
    service: BlockEmbedService
    size?: BlotSize
    title?: string
    type?: string
    uuid?: string
    version?: number
}

export interface ResizableEmbed {
    _renderSmall: Function
    _renderFullSize: Function
    handleModalOpen: (e: React.MouseEvent<HTMLElement, MouseEvent>) => void
    handleModalClose: (e: React.MouseEvent<HTMLElement, MouseEvent>) => void
}
