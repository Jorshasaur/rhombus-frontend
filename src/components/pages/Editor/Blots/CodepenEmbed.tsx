import * as React from 'react'
import { BlockEmbedProps } from '../../../../interfaces/blockEmbed'
import styles from './Blots.module.css'
import PrototypePlaceholderImage from '../../../../assets/images/embeds/prototype-placeholder.png'
import IconExpand from '../../../../assets/images/icons/icon-expand.svg'
import { Timestamp } from './Timestamp'
import { PersistentBar } from './PersistentBar'
import cx from 'classnames'
import { Button } from '@invisionapp/helios'
import { EmbedModalProps } from './EmbedModal'
import {
    createNewTabTracker,
    createModalOpenHandler,
    createModalCloseHandler
} from './BlotUtils'

export default class CodepenEmbed extends React.Component<BlockEmbedProps> {
    private trackOpenInNewTab = createNewTabTracker(this.props.service)

    private getIframeLink() {
        const { originalLink = '' } = this.props
        return originalLink.replace(/\/pen\//i, '/embed/')
    }

    getEmbedModalData = (): EmbedModalProps => ({
        onHide: this.handleModalClose,
        actionArea: (
            <a
                className={`${styles.newTabLink} styled-link`}
                data-testid="codepen-embed__new-tab"
                href={this.props.originalLink}
                onClick={this.trackOpenInNewTab}
                target="_blank">
                Open in New Tab...
            </a>
        ),
        children: (
            <iframe src={this.getIframeLink()} id="codepen-modal__iframe" />
        )
    })

    handleModalOpen = createModalOpenHandler(this.getEmbedModalData)

    handleModalClose = createModalCloseHandler()

    _renderFullSize = () => {
        return (
            <React.Fragment>
                <div
                    className={`
                        ${styles.blockEmbedWrapper}
                        ${this.props.service}
                        ${this.props.uuid}
                    `}>
                    <div
                        className={styles.extendedInfo}
                        style={{ borderTop: 0 }}>
                        <img
                            className={styles.ratio}
                            src={PrototypePlaceholderImage}
                        />
                        <div
                            className={cx(
                                styles.infoWrapper,
                                styles.thumbnailEmbed
                            )}>
                            <iframe src={this.getIframeLink()} />
                        </div>
                    </div>
                    <Button
                        data-testid="codepen-embed__expand-button"
                        data-allow-propagation="true"
                        order="primary"
                        className={styles.editButton}
                        // @ts-ignore
                        onMouseDown={(e: MouseEvent) => {
                            // Needed to prevent embed from being unselected while clicking button
                            e.preventDefault()
                        }}
                        // @ts-ignore
                        onClick={this.handleModalOpen}
                        style={{ padding: '1px 12px', borderRadius: '50%' }}>
                        <IconExpand
                            fill="white"
                            data-allow-propagation="true"
                            style={{ margin: 0 }}
                        />
                    </Button>
                </div>
                <PersistentBar
                    service={this.props.service}
                    onClick={this.handleModalOpen}
                    originalLink={this.props.originalLink}
                    name={this.props.title}
                    hoverChildren={
                        <div style={{ display: 'flex', alignItems: 'center' }}>
                            <Timestamp
                                author={this.props.authorName}
                                className={styles.persistentTimestamp}
                                createdAt={this.props.createdAt}
                            />
                        </div>
                    }
                />
            </React.Fragment>
        )
    }

    render() {
        return this._renderFullSize()
    }
}
