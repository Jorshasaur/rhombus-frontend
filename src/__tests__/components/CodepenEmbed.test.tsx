import React from 'react'
import { mount, shallow } from 'enzyme'
import CodepenEmbed from '../../components/pages/Editor/Blots/CodepenEmbed'
import { BlockEmbedService } from '../../interfaces/blockEmbed'
import { setupQlEditor, createContainer } from '../utils'
import { BlotSize } from '../../interfaces/blotSize'
import { ThemeProvider } from 'styled-components'
import theme from '@invisionapp/helios/css/theme'
import { EmbedModal } from '../../components/pages/Editor/Blots/EmbedModal'
import AnalyticsBuilder from '../../analytics/AnalyticsBuilders/AnalyticsBuilder'

const service: BlockEmbedService = 'codepen'

jest.mock('../../data/store', () => {
    return {
        subscribe: jest.fn(),
        getState: jest.fn(() => ({
            permissions: {
                canEdit: true
            }
        }))
    }
})

const Component = ({ children }: { children: React.ReactNode }) => (
    <ThemeProvider theme={theme}>{children}</ThemeProvider>
)

const baseProps = {
    createdAt: '2018-11-14T20:16:45.000Z',
    originalLink: 'https://codepen.io/FilipVitas/pen/yWyvwy',
    authorId: '1',
    authorName: 'Bob the Builder',
    key: '123456',
    service,
    uuid: '123456',
    version: 1,
    embedData: { id: '123' },
    container: createContainer(100, 100),
    size: BlotSize.Medium
}

jest.useFakeTimers()

describe('<CodepenEmbed />', () => {
    beforeAll(() => {
        setupQlEditor()
    })

    beforeEach(() => {
        EmbedModal.mount()
    })

    afterEach(() => {
        EmbedModal.unmount()
    })

    it('should render a Codepen embed', async () => {
        const wrapper = shallow(
            <Component>
                <CodepenEmbed {...baseProps} />
            </Component>
        )

        const codepen = wrapper.find('CodepenEmbed').dive()
        expect(codepen).toMatchSnapshot()
    })

    it('should set a timestamp', async () => {
        const wrapper = mount(
            <Component>
                <CodepenEmbed {...baseProps} />
            </Component>
        )
        const timestamp = wrapper.find('div.persistentTimestamp').first()
        expect(timestamp).toIncludeText('Added')
        expect(timestamp).toIncludeText(`ago by ${baseProps.authorName}`)
        const title = wrapper.find(testid('persistent-bar-title')).first()
        expect(title.text()).toContain('Codepen')
    })

    it('should open the modal', async () => {
        const track = jest.fn()
        AnalyticsBuilder.prototype.track = track

        const wrapper = mount(
            <Component>
                <CodepenEmbed {...baseProps} />
            </Component>
        )

        wrapper
            .find('Button' + testid('codepen-embed__expand-button'))
            .simulate('click')
        expect(EmbedModal.isShown()).toBe(true)
        expect(track).toHaveBeenCalled()
        document.querySelector<HTMLAnchorElement>('.newTabLink')!.click()
        expect(track).toHaveBeenCalledTimes(2)
        document
            .querySelector<HTMLAnchorElement>(
                testid('embed-modal__close-button')
            )!
            .click()
        expect(EmbedModal.isShown()).toBe(false)
        expect(track).toHaveBeenCalledTimes(3)
    })
})
