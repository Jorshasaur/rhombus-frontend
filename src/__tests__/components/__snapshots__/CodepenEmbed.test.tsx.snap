// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CodepenEmbed /> should render a Codepen embed 1`] = `
<Fragment>
  <div
    className="
                        blockEmbedWrapper
                        codepen
                        123456
                    "
  >
    <div
      className="extendedInfo"
      style={
        Object {
          "borderTop": 0,
        }
      }
    >
      <img
        className="ratio"
        src="prototype-placeholder.png"
      />
      <div
        className="infoWrapper thumbnailEmbed"
      >
        <iframe
          src="https://codepen.io/FilipVitas/embed/yWyvwy"
        />
      </div>
    </div>
    <Button
      className="editButton"
      data-allow-propagation="true"
      data-testid="codepen-embed__expand-button"
      onClick={[Function]}
      onMouseDown={[Function]}
      order="primary"
      role="button"
      style={
        Object {
          "borderRadius": "50%",
          "padding": "1px 12px",
        }
      }
    >
      <svg
        data-allow-propagation="true"
        fill="white"
        style={
          Object {
            "margin": 0,
          }
        }
      />
    </Button>
  </div>
  <PersistentBar
    hoverChildren={
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
          }
        }
      >
        <Unknown
          author="Bob the Builder"
          className="persistentTimestamp"
          createdAt="2018-11-14T20:16:45.000Z"
        />
      </div>
    }
    onClick={[Function]}
    originalLink="https://codepen.io/FilipVitas/pen/yWyvwy"
    service="codepen"
  />
</Fragment>
`;
